File: Signiture-Backend/kotlin/delivery/api/src/main/java/pcy/study/api/config/rabbitmq/RabbitMqConfig.java
package pcy.study.api.config.rabbitmq;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.amqp.core.Binding;
import org.springframework.amqp.core.BindingBuilder;
import org.springframework.amqp.core.DirectExchange;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.amqp.support.converter.MessageConverter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitMqConfig {

    private final String exchange;
    private final String queue;
    private final String routeKey;

    public RabbitMqConfig(
            @Value("${rabbitmq.exchange}")
            String exchange,
            @Value("${rabbitmq.queue}")
            String queue,
            @Value("${rabbitmq.route-key}")
            String routeKey
    ) {
        this.exchange = exchange;
        this.queue = queue;
        this.routeKey = routeKey;
    }

    @Bean
    public DirectExchange directExchange() {
        return new DirectExchange(exchange);
    }

    @Bean
    public Queue queue() {
        return new Queue(queue);
    }

    @Bean
    public Binding binding(DirectExchange exchange, Queue queue) {
        return BindingBuilder
                .bind(queue)
                .to(exchange)
                .with(routeKey);
    }

    @Bean
    public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory, MessageConverter messageConverter) {
        var rabbitTemplate = new RabbitTemplate(connectionFactory);
        rabbitTemplate.setMessageConverter(messageConverter);
        return rabbitTemplate;
    }

    @Bean
    public MessageConverter messageConverter(ObjectMapper objectMapper) {
        return new Jackson2JsonMessageConverter(objectMapper);
    }
}

File: Signiture-Backend/kotlin/delivery/api/src/main/kotlin/pcy/study/api/config/rabbitmq/RabbitMqConfig.kt
package pcy.study.api.config.rabbitmq

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.amqp.core.Binding
import org.springframework.amqp.core.BindingBuilder
import org.springframework.amqp.core.DirectExchange
import org.springframework.amqp.core.Queue
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitMqConfig(
    @Value("\${rabbitmq.exchange}")
    val exchange: String,
    @Value("\${rabbitmq.queue}")
    val queue: String,
    @Value("\${rabbitmq.route-key}")
    val routeKey: String,
) {

    @Bean
    fun directExchange(): DirectExchange {
        return DirectExchange(exchange)
    }

    @Bean
    fun queue(): Queue {
        return Queue(queue)
    }

    @Bean
    fun binding(
        exchange: DirectExchange,
        queue: Queue
    ): Binding {
        return BindingBuilder
            .bind(queue)
            .to(exchange)
            .with(routeKey)
    }

    @Bean
    fun rabbitTemplate(
        connectionFactory: ConnectionFactory,
        messageConverter: MessageConverter
    ): RabbitTemplate {
        return RabbitTemplate().apply {
            setConnectionFactory(connectionFactory)
            setMessageConverter(messageConverter)
        }
    }

    @Bean
    fun messageConverter(
        objectMapper: ObjectMapper
    ): MessageConverter {
        return Jackson2JsonMessageConverter(objectMapper)
    }
}
